### ISSUE TITLE
Consider exposing Client::BeginTransaction

### ISSUE THREAD

**Issue by coryan** [googleapis/google-cloud-cpp-spanner#627]
Think about an application that just wants to commit a few mutations, this does not work (safely) for retries:

```C++
auto txn = spanner::MakeReadWriteTransaction();
client.Commit(txn, my_mutations);
```

The problem is that this commit may succeed multiple times.  This does not work either, because it is the same thing:

```C++
RunTransaction(client, [](auto client, auto txn) {
client.Commit(txn, my_mutations);
});
```

And the [documentation](https://github.com/googleapis/googleapis/blob/dc88acb2ced136067d3ed2356b6a87ca7f017a96/google/spanner/v1/spanner.proto#L804-L813) reads:

```C++
    // ... Note that unlike
    // commit of a previously-started transaction, commit with a
    // temporary transaction is non-idempotent. 
```

**Comment by devjgm** [comment]
Rather than exposing `Client::BeginTransaction` to callers, I think we can solve this by having the implementation of `Commit` notice that a `Transaction` without a transaction ID is being used, then internally the *implementation* can call `stub_->BeginTransaction`, followed by the call to `stub_->Commit`.

I haven't thought this all the way through, but I think this may work and it would keep the user-facing API simpler.

Thoughts?

**Comment by coryan** [comment]
That would work.  I am not sure adding a simple operation like `BeginTransaction` complicates that user-facing API that much though `¯\_(ツ)_/¯`

**Comment by devjgm** [comment]
While it is a simple operation, the API complexity comes when we have to now explain to users when to call it, what it does, and that other `Transaction` instances may or may not actually exist with an ID (that the user can't see) until they call either `BeginTransaction` or call some other method. I think it'd be good if we can keep that complexity out of the API.

**Comment by coryan** [comment]
Well, now `Commit()` either makes 2 RPCs or 1 RPC depending on whether the transaction has an id already or not.  Or `Commit()` is idempotent (and retried) depending on whether it has an id or not.

I am fine with not introducing `BeginTransaction()`, I do not think it is that important to add it or to keep it out.  I am just pointing out how the actual behavior is starting to leak.

**Comment by devjgm** [comment]
Why is it a problem if Commit() (maybe) makes 2 RPCs? I don't see what that is leaking.

**Comment by devjgm** [comment]
I'm thinking we can/should close this issue. But let me first make sure @devbww and @mr-salty saw this, since I believe this is related to stuff that we've considered before.

**Comment by coryan** [comment]
Closing this is Okay by me.

For the record, 2 RPCs are observable because the performance is going to be very different vs. 1 RPC.

**Comment by mr-salty** [comment]
I think we can close it. FWIW, Session Pooling will mean that ideally this is still 1 RPC regardless because we've "prepared" the session by doing `BeginTransaction` already.

Carlos' comment about "leaking" is interesting - I'd say we're doing the opposite (hiding the detail of requiring `BeginTransaction`) but hiding details also hides some performance implications (I don't think I'd call that "leaking" though)

**Comment by devbww** [comment]
Perhaps we should be a little more specific with our conclusions before closing.

One situation we discussed some time back was a large set of parallel reads, where we currently serialize the first read (to bind the ID) before allowing the rest to actually proceed in parallel.  The thought was that an initial `BeginTransaction()` followed by N parallel reads might be faster than that first read followed by N-1 parallel reads.  It's unclear what the tradeoffs are there however, so I was certainly happy to forego `BeginTransaction()` until someone raised the question.

In this issue, Carlos raises the situation of a `Commit()` with an unbound ID not being idempotent, and therefore not being retry-able in cases where the request may have reached the server.  Greg suggests never making a `Commit()` call with an unbound ID by doing a `BeginTransaction()` behind the scenes, which then makes it retry-able.  I'm still unclear on what a retry-able `Commit()` even means though.  For example, if we lose the first (successful) response to a `Commit()`, and then retry, do we get some form of "ID already committed" response that allows us to declare success to the user?  Is a non-local failure to `Commit()` ever even retry-able?

I don't think the "one RPC" -v- "two RPC" question is important.  If we believe it is important to `BeginTransaction()` before `Commit()`, it is two RPCs whether it is one user call or two.  One user call sounds simpler.

And Todd's observation that we still might not need a user-level `BeginTransaction()` if we have a transaction ID always(?) ready to go for a session, muddies the water further.  I'm not even sure how such a facility would work though.

Anyway, my point is that I think this could do with further discussion.

**Comment by devjgm** [comment]
Discussion about which question? If I'm reading correctly, it sounds like all conclusions point to not exposing a user-level `BeginTransaction` call. Is that correct? If so, I think we should close this issue. If there's another question that we need to discuss and answer, maybe we should file a separate issue for that?

Sorry if I've misunderstood something here.

**Comment by mr-salty** [comment]
on parallel reads possibly needing `BeginTransaction`, I'm not sure if that is a case we need to worry about, or if you should be using partitioned queries or similar to accomplish that - spanner docs [kind of suggest so](http://cloud/spanner/docs/reads#read_data_in_parallel) - so let's disregard that that for now.

> In this issue, Carlos raises the situation of a `Commit()` with an unbound ID not being idempotent, and therefore not being retry-able in cases where the request may have reached the server. Greg suggests never making a `Commit()` call with an unbound ID by doing a `BeginTransaction()` behind the scenes, which then makes it retry-able. I'm still unclear on what a retry-able `Commit()` even means though. For example, if we lose the first (successful) response to a `Commit()`, and then retry, do we get some form of "ID already committed" response that allows us to declare success to the user? Is a non-local failure to `Commit()` ever even retry-able?

I think this is a pretty big yet subtle issue, but doesn't (necessarily) affect our decision about exposing `BeginTransaction`. I opened #689

> I don't think the "one RPC" -v- "two RPC" question is important. If we believe it is important to `BeginTransaction()` before `Commit()`, it is two RPCs whether it is one user call or two. One user call sounds simpler.

I agree, but let's also discuss that in #689 since the 2-call performance implication is a potential downside of what I proposed there.

> And Todd's observation that we still might not need a user-level `BeginTransaction()` if we have a transaction ID always(?) ready to go for a session, muddies the water further. I'm not even sure how such a facility would work though.

Note this is only for read-write, since there are no read-write options we can always call `BeginTransaction`.  Doing this assumes we will never have any RW options, or that they're specific to a `Client` or similar instead of per-transaction, and it also means we may need to remove `ReadWriteOptions` from our `MakeReadWriteTransaction` API.  The design for this is currently internal (and not finished) so we should discuss offline.

**Comment by devjgm** [comment]
We discussed this offline. We are not going to expose `Client::BeginTransaction`. The remaining question about what `Client::Commit` does is discussed in #689.
